### PR TITLE
Document all undocumented xkit_patches functions

### PR DESCRIPTION
**Pull request by hobinjk**
Note that there is no version bump because there are no changes. Also documented during this are the functions that xkit_patches.js overrides from xkit.js, which are:
> XKit.tools.get_blogs
> XKit.tools.add_function
> XKit.browser
> XKit.window.close
> XKit.window.show
> XKit.init
> XKit.post_listener.check
> XKit.post_listener.run_callbacks
> XKit.notifications.add

### CODE REVIEW

File: Extensions/xkit_patches.js [hunk 1]
```diff
@@ -331,44 +389,73 @@ XKit.tools.getParameterByName = function(name){
...
+			 *                  the blog in which the iframe is embedded)
+			 */
 			get_tumblelog: function() {
 				var new_channel_id = document.location.href.match(/[&?]tumblelogName=([\w-]+)/);
 				var old_channel_id = $("#tumblelog_name").attr('data-tumblelog-name');
 				return (new_channel_id && new_channel_id[1]) || old_channel_id;
 			},
 
+			/**
+			 * @return {String?} Post which this iframe refers to or undefined
```
Comment (nightpool) [inline Extensions/xkit_patches.js]: cannot return undefined—will throw a TypeError if not run in an iframe. (this is normally never an issue, as all users currently use other methods to determine if they're in a post or not, but i'm down for changing if you want to)

File: Extensions/xkit_patches.js [hunk 2]
```diff
@@ -331,44 +389,73 @@ XKit.tools.getParameterByName = function(name){
...
+			 *                  reblog or other API request)
+			 */
 			form_key: function() {
 				var new_form_key = $("meta[name=tumblr-form-key]").attr("content");
 				var old_form_key = $(".btn.reblog").attr('data-form-key');
 		        return new_form_key || old_form_key;
 			},
 
+			/**
+			 * @param {jQueryElement} button - The button in the iframe to be hidden
```
Comment (nightpool) [inline Extensions/xkit_patches.js]: name is misleading—this function just hides the text of the button, not the icon

File: Extensions/xkit_patches.js [hunk 3]
```diff
@@ -331,44 +389,73 @@ XKit.tools.getParameterByName = function(name){
...
+			 */
 			hide_button: function(button) {
 				button.addClass("no_label");
 				button.addClass("no-text").wrapInner('<span class="hidden">');
 			},
 
+			/**
+			 * @return {jQueryElement} The follow button in the iframe
```
Comment (nightpool) [inline Extensions/xkit_patches.js]: isn't this a list of jquery elements? or something? a jQuery selector? not sure what types jquery has, exactly.

File: Extensions/xkit_patches.js [hunk 4]
```diff
@@ -236,6 +282,17 @@ XKit.tools.getParameterByName = function(name){
 	}
 };
 
+/**
+ * @return {Object} An overview of the browser's information:
+ *	name<"Google Chrome" | "Mozilla Firefox" | "Apple Safari">: The browser's human-readable name
```
Comment (nightpool) [inline Extensions/xkit_patches.js]: is there a standard you're using for these complex type definitions? if so, that's fine, but if not I would prefer a slightly different format

File: Extensions/xkit_patches.js [hunk 5]
```diff
@@ -201,6 +232,13 @@ XKit.tools.make_gist = function(text, name) {
 	});
 };
 
+/**
+ * Injects a function into the DOM scope.
```
Comment (nightpool) [inline Extensions/xkit_patches.js]: I'm not sure if "DOM scope" is a very precise name for what we're doing here, maybe "page context" would be better? can't remember offhand how its referred to.

alternative wording:

> Copies a function from the addon context into the page context. This function will be serialized to a string, and then injected as a script tag into the page. `addt` will be serialized as JSON and then passed as a parameter to the function

File: Extensions/xkit_patches.js (hunk 1 above)
Comment (hobinjk) [inline Extensions/xkit_patches.js]: Good catch, I forgot that `match` returns `null` instead of `[]`

File: Extensions/xkit_patches.js (hunk 4 above)
Comment (hobinjk) [inline Extensions/xkit_patches.js]: The string part is from Typescript's String Literal Types specification as described in https://github.com/Microsoft/TypeScript/wiki/What's-new-in-TypeScript#string-literal-types.

The description of keys using parameterization is something I stopped doing later and I will update this location to the one I eventually decided on which was "name: Type - description"

What is your preference?

File: Extensions/xkit_patches.js (hunk 3 above)
Comment (hobinjk) [inline Extensions/xkit_patches.js]: I've been using jQueryElement as the return type of `$(anything)`. jQuerySelector might be slightly clearer. TypeScript uses `JQuery` as their return type, which I think I'll use for consistency unless you have a different preference.

File: Extensions/xkit_patches.js (hunk 5 above)
Comment (hobinjk) [inline Extensions/xkit_patches.js]: :+1:, will update with your eloquence

Comment (hobinjk) [inline Extensions/xkit_patches.js]: Although `addt` is not actually passed as a parameter. Will update with parts of your eloquence.

File: Extensions/xkit_patches.js (hunk 4 above)
Comment (nightpool) [inline Extensions/xkit_patches.js]: I like name: Type - description

File: Extensions/xkit_patches.js (hunk 3 above)
Comment (nightpool) [inline Extensions/xkit_patches.js]: down for whatever. element is just a little unclear—its certainly some sort of plural collection type. using jQuery is a good compromise.

File: Extensions/xkit_patches.js [hunk 6]
```diff
@@ -414,6 +507,16 @@ XKit.tools.getParameterByName = function(name){
 
 		};
 
+		/**
+		 * Show an XKit alert window
+		 * @param {String} title - Text for alert window's title bar
+		 * @param {String} msg - Text for body of window, can be HTML
+		 * @param {"error"|others} icon - Window's icon type, determined by CSS class `icon`.
```
Comment (nightpool) [inline Extensions/xkit_patches.js]: probably should either put all the possible values in here or make a constant for it or something.

File: Extensions/xkit_patches.js [hunk 7]
```diff
@@ -1255,9 +1464,13 @@ XKit.tools.getParameterByName = function(name){
 
 			},
 
+			/**
+			 * Switches control button between "working" and normal.
```
Comment (nightpool) [inline Extensions/xkit_patches.js]: could be a little clearer

>Switches control button between normal and the "working" animation

File: Extensions/xkit_patches.js [hunk 8]
```diff
@@ -1268,9 +1481,13 @@ XKit.tools.getParameterByName = function(name){
 
 			},
 
+			/**
+			 * Switches control button between "disabled" and normal.
```
Comment (nightpool) [inline Extensions/xkit_patches.js]: "switches" implies that this is a toggle, which it isn't—it depends on the passed value.

File: Extensions/xkit_patches.js [hunk 9]
```diff
@@ -1294,6 +1515,16 @@ XKit.tools.getParameterByName = function(name){
 
 			},
 
+			/**
+			 * Specify a control button that can be added to future posts
```
Comment (nightpool) [inline Extensions/xkit_patches.js]: "specify" is a weird verb for a constructer.

File: Extensions/xkit_patches.js [hunk 10]
```diff
@@ -1294,6 +1515,16 @@ XKit.tools.getParameterByName = function(name){
...
+			 * using `XKit.interface.add_control_button`.
+			 * @param {String} class_name - CSS class of the button to be created
+			 * @param {String} icon - URL of the button's icon
+			 * @param {String} text - Hover text of the button
+			 * @param {Function<Event>} func - Function called on click of control button
```
Comment (nightpool) [inline Extensions/xkit_patches.js]: should probably be something like "EventHandler" -- function<event> is ambiguous to whether its parameterizing the return type or the arguments

Comment (nightpool) [inline Extensions/xkit_patches.js]: mdn calls this an [EventListener](https://developer.mozilla.org/en-US/docs/Web/API/EventListener)

File: Extensions/xkit_patches.js [hunk 11]
```diff
@@ -1344,6 +1575,13 @@ XKit.tools.getParameterByName = function(name){
 
 			},
 
+			/**
+			 * Instantiate and add a previously "created" button to the
+			 * specified post.
+			 * @param {Object} obj - Interface Post Object
+			 * @param {String} class_name - CSS class of the button to be added
+			 * @param {String?} additional - Additional parameters for the button's div
```
Comment (nightpool) [inline Extensions/xkit_patches.js]: unclear. Should mention that this is in the form of a string that get inserted into the button's tag.

File: Extensions/xkit_patches.js (hunk 9 above)
Comment (hobinjk) [inline Extensions/xkit_patches.js]: I updated this and its later use to be "Create a specification for a control button [...]". The most important idea that I'm trying to get across is that it doesn't instantiate anything or actually create the button (despite its name).

File: Extensions/xkit_patches.js (hunk 10 above)
Comment (hobinjk) [inline Extensions/xkit_patches.js]: I've updated all uses of Function<Event> to be EventListener and all uses of Function<T> to be Function with information about expected arguments.

**Comment by hobinjk** [comment]
Updated with full enumeration of icon type values

**Comment by hobinjk** [comment]
Merge conflicts fixed. I believe all review comments are also fixed.

**Comment by nightpool** [comment]
this came up when I was working on a new tools function: should be also be maintaining the documentation under the docs/ folder? Are we going to be auto-generating stuff there?

**Comment by hobinjk** [comment]
My goal is to auto-generate as much as possible from JSDoc to make sure that effort isn't duplicated.

**Comment by nightpool** [comment]
Okay, looks good to me then

@homu r+

**Pull request merged**